Wikipedia/Code/WMFAuthTokenFetcher.swift

```diff
@@ -67,9 +78,48 @@ public class WMFAuthTokenFetcher: NSObject {
                 failure(WMFAuthTokenError.zeroLengthToken)
                 return
             }
+            guard token != Constants.anonymousToken else {
+                if self.shouldRetryWithLogin {
+                    //DDLogDebug("Fetched anonymous token for \(siteURL), retrying with login")
+                    let context = FetchTokenContext(tokenType: type, siteURL: siteURL, success: success, failure: failure)
+                    self.retryWithLogin(context: context)
+                } else {
+                    success(WMFAuthToken(token: token, type: type))
+                    return
+                }
+                return
+            }
             success(WMFAuthToken.init(token: token, type: type))
         }, failure: { (_, error) in
             failure(error)
         })
     }
+
+    private struct FetchTokenContext {
+        let tokenType: WMFAuthTokenType
+        let siteURL: URL
+        let success: WMFAuthTokenBlock
+        let failure: WMFErrorHandler
+    }
+
+    private func retryWithLogin(context: FetchTokenContext) {
+        guard !isRetryingLogin else {
+            //DDLogDebug("Still retrying with login; aborting another attempt")
+            return
+        }
+        isRetryingLogin = true
+        let fetchToken = {
+            self.fetchToken(ofType: context.tokenType, siteURL: context.siteURL, success: context.success, failure: context.failure)
+        }
+        WMFAuthenticationManager.sharedInstance.loginWithSavedCredentials(context.siteURL, success: { (_) in
+            fetchToken()
+        }, userAlreadyLoggedInHandler: { (_) in
+            fetchToken()
+        }, failure: { (error) in
+            context.failure(error)
+        }) {
+            self.isRetryingLogin = false
```

Comment:
There might need to be some assurance that this is thread safe - maybe a main thread assert above and an async to the main thread here or a lock or an access queue